src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,1070 @@
+    // cannot be, because it must be 0. As a result, the last bit we can flip,
+    // which is the first different bit between the result and lo must be the
+    // 3rd bit. As a result, the result must not be smaller than:
+    //           1 0 1 0 0 0 0 0
+    // This one satisfies zeros so we can use the logic in the previous case,
+    // just OR with ones to obtain the final result, which is:
+    //           1 0 1 0 0 0 1 1
+
+    juint first_violation = count_leading_zeros(zero_violation);
+    // This masks out all bits after the first violation
+    //           1 1 1 1 1 0 0 0
+    U find_mask = ~(std::numeric_limits<U>::max() >> first_violation);
+    // A bit that is 0 in both lo and zeros must be 0 in either
+    //           1 0 0 1 1 1 1 0
+    U either = lo | bits._zeros;
+    // This is all the bits that are not lower than first_violation and are 0
+    // in both lo and zeros. Note that there may not exist such bits. In those
+    // cases tmp == 0 and there is no value not less than lo and satisfies
+    // bits. We will expand the implication of such cases below.
+    //           0 1 1 0 0 0 0 0
+    U tmp = ~either & find_mask;
+    // According to the conclusion in the overview of this case above, i is the
+    // last bit being 0 in both lo and zeros that is upto the first violation,
+    // which is the last set bit of tmp.
+    // In our example, i == 2 here, we want to obtain the value with only the
+    // bit i set, this is equivalent to extracting the last set bit of tmp, do
+    // it directly without going through i.
+    //           0 0 1 0 0 0 0 0
+    U alignment = tmp & (-tmp);
+    // Set the bit of lo at i and unset all the bits after, this is the smallest
+    // value that satisfies bits._zeros. Similar to the above case, this is
+    // similar to aligning lo upto alignment. Also similar to the above case,
+    // this computation cannot overflow.
+    // We now have:
+    // - new_lo[x] = lo[x], for 0 <= x < i (2.5)
+    // - new_lo[i] = 1                     (2.6)
+    // - new_lo[x] = 0, for x > i          (not yet 2.7)
+    //           1 0 1 0 0 0 0 0
+    U new_lo = (lo & -alignment) + alignment;
+    // Note that there exists no value x not larger than i such that
+    // new_lo[x] == 0 and ones[x] == 1. This is because all bits of lo before i
+    // should satisfy bits, and new_lo[i] == 1. As a result, doing
+    // new_lo |= bits.ones will give us a value such that:
+    // - new_lo[x] = lo[x], for 0 <= x < i (2.5)
+    // - new_lo[i] = 1                     (2.6)
+    // - new_lo[x] = ones[x], for x > i    (2.7)
+    // This is the result we are looking for.
+    //           1 0 1 0 0 0 1 1
+    new_lo |= bits._ones;
+    // In this case, new_lo may not always be a valid answer. This can happen
+    // if there is no bit upto first_violation that is 0 in both lo and zeros,
+    // i.e. tmp == 0. In such cases, alignment == 0 && lo == bits._ones. It is
```

Comment:
```suggestion
    // i.e. tmp == 0. In such cases, alignment == 0 and new_lo == bits._ones. It is
```
Or do we somehow also know that `lo = ones`? I think this was a typo, right?